[Replicate][Test] tests that pp model grads are the same as single-device model grads

anshul-si opened the pull request:
**Summary:** Created a test so that we can verify that a model that has been pipelined + replicated has the same gradients as a reference model. To do this, I mapped the layers and their parameters in each partial model to the original full model and then compared the gradients. 
**Test Case**
1. pytest test/distributed/_composable/test_composability/test_pp_composability.py -k test_replicate_pp_grads

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #164890



cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

anshul-si committed:
[Replicate][Test] tests that pp model grads are the same as single-device model grads

[ghstack-poisoned]

H-Huang commented on test/distributed/_composable/test_composability/test_pp_composability.py:
```diff
@@ -604,6 +604,284 @@ def apply_same_precision(partial_model):
+                optimizer.step()
+            ref_optimizer.step()
+
+            # torch.distributed.breakpoint()
```
remove

H-Huang reviewed (approved):
LGTM, just some small comments / questions. Feel free to land it when you are ready!

anshul-si committed:
Update on "[Replicate][Test] tests that pp model grads are the same as single-device model grads"

**Summary:** Created a test so that we can verify that a model that has been pipelined + replicated has the same gradients as a reference model. To do this, I mapped the layers and their parameters in each partial model to the original full model and then compared the gradients. 
**Test Case**
1. pytest test/distributed/_composable/test_composability/test_pp_composability.py -k test_replicate_pp_grads




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]
